### PR TITLE
Add the secp256k1 and secq256k1 curves

### PR DESCRIPTION
**Pull request by weikengchen**
## Description

Relies on https://github.com/arkworks-rs/algebra/pull/469. When that one is merged, we will need to update the Cargo.toml to point to the `master` branch of algebra.

A change is needed to fix the constraint tests.

Closes: #36 
It also makes the branch `secp256k1` obsolete. @npwardberkeley 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

N/A:
- [ ] Wrote unit tests
- [ ] Updated relevant documentation in the code

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -60,7 +62,7 @@ debug = true
 # To be removed in the new release.
 [patch.crates-io]
 ark-ec = { git = "https://github.com/arkworks-rs/algebra" }
-ark-ff = { git = "https://github.com/arkworks-rs/algebra" }
+ark-ff = { git = "https://github.com/arkworks-rs/algebra"}
```
Comment (weikengchen) [inline Cargo.toml:65]:
```suggestion
ark-ff = { git = "https://github.com/arkworks-rs/algebra" }
```

File: curve-constraint-tests/src/lib.rs [hunk 2]
```diff
@@ -320,9 +320,15 @@ pub mod curves {
             for limb in &mut max {
                 *limb = u64::MAX;
             }
-
             let modulus_last_limb_bits = <C::ScalarField as PrimeField>::MODULUS_BIT_SIZE % 64;
```
Comment (weikengchen) [inline curve-constraint-tests/src/lib.rs:323]:
```suggestion

            let modulus_last_limb_bits = <C::ScalarField as PrimeField>::MODULUS_BIT_SIZE % 64;
```

**Comment by Pratyush** [comment]
Instead of implementing it on BigInt320, how about we port the techniques from [the crrl library](https://github.com/pornin/crrl/blob/3a3e625d784b4234d53829771fc3e18abab16913/src/backend/w64/modint.rs#L14)? That avoids allocating an extra limb.

Also, we can port the specialized arithmetic from the base field over.

**Comment by weikengchen** [comment]
@Pratyush Let me create an issue to cover this topic, and probably for this PR we will just let it be this.

**Comment by weikengchen** [comment]
This PR is ready.

File: secp256k1/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    future_incompatible,
+    nonstandard_style,
+    rust_2018_idioms
+)]
+#![forbid(unsafe_code)]
+
+//! This library implements the secp256k1 curve.
+//!
+//! Curve information:
+//! Vesta:
```
Comment (Pratyush) [inline secp256k1/src/lib.rs:14]:
```suggestion
```

File: secq256k1/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    future_incompatible,
+    nonstandard_style,
+    rust_2018_idioms
+)]
+#![forbid(unsafe_code)]
+
+//! This library implements the secq256k1 curve.
+//!
+//! Curve information:
+//! Vesta:
```
Comment (Pratyush) [inline secq256k1/src/lib.rs:14]:
```suggestion
```

File: secq256k1/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+#![deny(
+    warnings,
+    unused,
+    future_incompatible,
+    nonstandard_style,
+    rust_2018_idioms
+)]
+#![forbid(unsafe_code)]
+
+//! This library implements the secq256k1 curve.
```
Comment (Pratyush) [inline secq256k1/src/lib.rs:11]: Could you add a reference for the definition of the curve?

File: secp256k1/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    warnings,
+    unused,
+    future_incompatible,
+    nonstandard_style,
+    rust_2018_idioms
+)]
+#![forbid(unsafe_code)]
+
+//! This library implements the secp256k1 curve.
+//!
```
Comment (Pratyush) [inline secp256k1/src/lib.rs:12]: Could you add a reference for the curve definition?

**Review by Pratyush — CHANGES_REQUESTED** [review]

**Comment by weikengchen** [comment]
Sources added.

File: secq256k1/src/lib.rs (hunk 5 above)
Comment (weikengchen) [reply secq256k1/src/lib.rs:11]: DONE

File: secp256k1/src/lib.rs (hunk 6 above)
Comment (weikengchen) [reply secp256k1/src/lib.rs:12]: DONE

Comment (Pratyush) [reply secp256k1/src/lib.rs:13]: Can you do this also for secq?

Comment (Pratyush) [reply secp256k1/src/lib.rs:13]: And by source, I meant a formal description of the curve constants, e.g. https://en.bitcoin.it/wiki/Secp256k1

Comment (Pratyush) [reply secp256k1/src/lib.rs:13]: Or something like this: https://www.secg.org/sec2-v2.pdf

**Comment by weikengchen** [comment]
Fixed. I incorrectly put the secq one in the secp. Now it is fixed. For secq I think that thread is the citation that we generally use when referring to secq.

File: secp256k1/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    warnings,
+    unused,
+    future_incompatible,
+    nonstandard_style,
+    rust_2018_idioms
+)]
+#![forbid(unsafe_code)]
+
+//! This library implements the secp256k1 curve.
+//! Source: <https://en.bitcoin.it/wiki/Secp256k1>
```
Comment (weikengchen) [inline secp256k1/src/lib.rs:12]: Fixed as here.

File: secq256k1/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+//! Source: <https://moderncrypto.org/mail-archive/curves/2018/000992.html>
```
Comment (weikengchen) [inline secq256k1/src/lib.rs:12]: Fixed as here.

**Review by Pratyush — APPROVED** [review]

**Comment by weikengchen** [comment]
I deleted the `secp256k1` branch as well. (@npwardberkeley Nick Ward's  previous attempt)

**Pull request merged**